### PR TITLE
Variable autobaud

### PR DESCRIPTION
**Pull request by hexfet**
The autobaud registers count edges and keep track of the shortest high and low pulses.  This information is used to calculate the bitrate.  This method makes it easier to support a variety of bitrates, adapts to individual handset differences, and is faster than trying a list of baudrates.  Original motivation was making it easier to test what rate a handset can support.

While the reference manual states the baud rate should be 80MHz / min(low pulse count, high pulse count) it seems that adding 2 to the count produces better results.  In the table the 2.25MHz datarate was tested on a T8SG+, and the others with a T16.  Radios were HappyModel ES2400TX and EP1 RX. The calculated rate is what I think the handsets run at based on the stm32f4 reference manual and edgetx code (84MHz pclk, 8x sampling for rates over 400k).

| handset baud rate | calculated rate | autobaud count | UARTrequestedBaud clk/count | error % | UARTrequestedBaud clk/(count+1) | error % | UARTrequestedBaud clk/(count+2) | error % |
|-------------------|------------|----------------|------------|----------------|------------|----------------|------------|----------------|
| 115200 | 115226 | 692 | 115,606 | 0.33% | 115,440 | 0.19% | 115,273 | 0.04% |
| 400000 | 400000 | 198* | 404,040 | 1.01% | 402,010 | 0.50% | 400,000 | 0.00% |
| 921600 | 923076 | 86 | 930,232 | 0.78% | 919,540 | -0.38% | 909,090 | -1.52% |
| 1870000 | 1866666 | 40 | 2,000,000 | 7.14% | 1,951,219 | 4.53% | 1,904,761 | 2.04% |
| 3750000 | 3818181 | 19 | 4,210,526 | 10.28% | 4,000,000 | 4.76% | 3,809,523 | -0.23% |
| 2250000 | 2250000 | 34 | 2,352,941 | 4.58% | 2,285,714 | 1.59% | 2,222,222 | -1.23% |

\* tested but unused as 400K is the default datarate

Update 23 Mar 2022: Added automatic uart inversion detection.  Helps users who swap between inverted and non-inverted handsets.

### CODE REVIEW

**Comment by hexfet** [comment]
I couldn't find an obvious way to pull the uart number from CRSF::Port in order to remove the hard-coded 0.  Same with the 80MHz clock rate.  The HAL has uart_ll_* functions for accessing the autobaud registers but I could not find them exposed further in the API.  Stumbled across the uart register macros but there may be a better way.

**Comment by JyeSmith** [comment]
Tagging @AlessandroAU because I know he has previous experience in this.

**Comment by AlessandroAU** [comment]
Hi @hexfet sorry for the delay, I've been busy with work. This is a great idea! We also had internal discussion about how to deal with baud error margin at higher UART speeds. I wonder if this could effectively solve that issue? 

The only concern I have is that as I understand it, autobaud relies on a specific training sequence of minimal and maximal length pulses. I don't know if it is possible to guarantee that those will occur in the data stream. 

@CapnBry @pkendall64 do you any thoughts?

**Comment by pkendall64** [comment]
> Hi @hexfet sorry for the delay, I've been busy with work. This is a great idea! We also had internal discussion about how to deal with baud error margin at higher UART speeds. I wonder if this could effectively solve that issue?
> 
> The only concern I have is that as I understand it, autobaud relies on a specific training sequence of minimal and maximal length pulses. I don't know if it is possible to guarantee that those will occur in the data stream.
> 
> @CapnBry @pkendall64 do you any thoughts?

Sorry I haven't got to this either, I've been busy with other ELRS things.

My understanding of the Espressif docs is that any random data seen on the UART can be used to detect the min width of a LOW pulse and HIGH pulse and from that you can calculate the baud rate.
So no specific training sequence is required, just so long as the data on UART does contain single instances of 0's and 1's so it can detect those minimums correctly.

**Comment by pkendall64** [comment]
While I like the idea of this, in practice it doesn't seem to work too well.
When I ran with the code as in the PR, 115k and 5.25M would just not work. I adjusted the offset to 1 from 2 and now 5.25M works, but 3.75 is not. I think that rather can calculating the baud rate to use the minimum measured interval could be used to automatically select the correct baud from the list of valid values, this way be get the best of both worlds. i.e. all baud rates are known and work, and we don't have to cycle through them, we can detect the current running rate.

File: src/lib/CRSF/CRSF.cpp [hunk 1]
```diff
@@ -764,6 +771,47 @@ void ICACHE_RAM_ATTR CRSF::adjustMaxPacketSize()
...
+    state = MEASURED;
+
+    uint32_t low_period  = *(uint32_t *)UART_LOWPULSE_REG(0);
+    uint32_t high_period = *(uint32_t *)UART_HIGHPULSE_REG(0);
+    *autobaud_reg = (4 << 8) | 0;
+
+    DBGLN("autobaud: low %d, high %d", low_period, high_period); 
+    // tech ref says baud rate = 80000000/min(UART_LOWPULSE_REG, UART_HIGHPULSE_REG);
+    // add 2 based on testing for lowest deviation
+    return 80000000 / (min(low_period, high_period) + 2);
```
Comment (pkendall64) [inline src/lib/CRSF/CRSF.cpp:806]: Changing this line to a closest match lookup works reliably
```cpp
    int32_t calulatedBaud = 80000000 / (min(low_period, high_period) + 2);
    int32_t bestBaud = (int32_t)TxToHandsetBauds[0];
    for(int i=0 ; i<ARRAY_SIZE(TxToHandsetBauds) ; i++)
    {
        if (abs(calulatedBaud - bestBaud) > abs(calulatedBaud - (int32_t)TxToHandsetBauds[i]))
        {
            bestBaud = (int32_t)TxToHandsetBauds[i];
        }
    }
    return bestBaud;
```

**Comment by hexfet** [comment]
No worries on the time - this stuff should be stress-free.

@pkendall64 Thanks for testing.  Running through all the rates on my T16 I found that 115k and 400k no longer work while the others were fine.  My fault for not running all tests since making the table.  The 400k result is puzzling since that's the default rate.  It works in the t8sg and I compiled with uart inversion off, which hints that maybe the added code for auto-inversion interferes with the process somehow.  I'm currently traveling so won't be able to spend much time investigating till next week.

I like the approach @pkendall64 suggests.  With that approach I would like to add 2.25M to the baud rate list to better support Deviation. 

On the question of ensuring the measurement integrity, the algorithm requires measuring minimum length of high and low pulses.  The handset will be sending mostly rc packets when the data is collected.  These start with EE1816, which results in 5 single-bit-width lows and 3 highs, plus the stick data should be somewhat random for used channels.  The edge count check at line 794 of CRSF.cpp ensures a minimum number of edges have been counted.  It occurs to me another check could be added to ensure the low and high counts are nearly the same.  In testing I never saw them differ by more than 1.

**Comment by pkendall64** [comment]
> No worries on the time - this stuff should be stress-free.
> 
> @pkendall64 Thanks for testing. Running through all the rates on my T16 I found that 115k and 400k no longer work while the others were fine. My fault for not running all tests since making the table. The 400k result is puzzling since that's the default rate. It works in the t8sg and I compiled with uart inversion off, which hints that maybe the added code for auto-inversion interferes with the process somehow. I'm currently traveling so won't be able to spend much time investigating till next week.
> 
> I like the approach @pkendall64 suggests. With that approach I would like to add 2.25M to the baud rate list to better support Deviation.
> 

I can push my changes if you're ok with that, and I can also add 2.25M to the list of supported buad rates as well.

**Comment by hexfet** [comment]
> I can push my changes if you're ok with that, and I can also add 2.25M to the list of supported buad rates as well.
Yes, I'm fine with that.  With your suggested growing the list should be fine as trying each rate is no longer necessary.  Thanks!

**Comment by hexfet** [comment]
Still looking into this.  Time's been limited and I'm traveling again this week.

The measurement part sometimes does not work correctly - the low and high counts do not match.  There are plenty of edges in the bitstream so not clear why high and low pulse widths don't always measure the same.  Usually the high value is correct and the low count is much lower.  Have some ideas on things to look at but no opportunity till next week.

**Comment by hexfet** [comment]
I've not been able to determine why the high and low pulse width counts from the autobaud registers are not the same.  I've updated the PR with one change based on many observations that the higher of the two values is always correct.  The test of which value to use to calculate the baud rate is changed from min() to max().

I've left in the "best match" code even though with the above change it does not seem necessary (and I can't test 5.25M). It does remove the ability to adjust to any handset baudrate, but without a solid explanation for the autobaud behavior it seems a good idea.  The autobaud code still eliminates the need to test multiple baudrates.

Testing notes
* At 115K and 400K the high and low pulse counts do not always match.  The larger value is always correct and is very stable.  The smaller value is not consistent but tends to vary in a small range with each measurement interval.  For example, successive measurements at 115K always produce a large value of 692 or 693, while the smaller varies (332, 312, 342, ...) or (142, 119, 137, ...).  The larger measurement is almost always the high pulse width, but not always.  Especially at 400K the larger number may be either the high or the low count.
* Packet rate has an effect on the measurements.  At 115K and 50HZ the two counts will never match and generally (but not always) the low value is in the 300's.  At 115K and 250Hz, usually after tens of measurements both high and low counts will measure 692/693.
* Trying to wait for serial line idle before starting the measurement didn't change the behavior.  Used the UART status register to check both that the state machine was in rx_idle and the RXD line was high before staring the autobaud measurement. Couldn't find state machine details so its behavior when baud rate is not correct is unknown.
* Checked the serial data stream for short pulses that might produce the lower count values, but did not find any.
* The code comment about the tech ref manual was not correct. Section 13.3.4 just says the counts can be used to calculate baudrate. Using the min() value came from code example [here](https://github.com/espressif/esp-idf/issues/3336).

**Review by pkendall64 — APPROVED** [review]

**Review by JyeSmith — APPROVED** [review]

**Pull request merged**